### ISSUE TITLE
`NotRequired` and `Required` not properly handled with Python < 3.11 in `TypedDict` conversion

### ISSUE THREAD

**Issue by gomarz** [robotframework/robotframework#5115]
Hi All, I have a few classes that use the NotRequired notation from typing_extensions_4.8.0 and python 3.9.13
```
class Signal(TypedDict):
    Network: str
    Signal: str

class WriteSignal(Signal):
    Value: str
    Interval: NotRequired[int]
```

This worked fine in RF 6 but when trying to upgrade my project to RF7, I have been encountering this error when running robot tests using the keyword that utilizes this class:

> [ ERROR ] Error in library 'RobotLibrary': Adding keyword 'CANWrite' failed: 'typing_extensions.NotRequired[int]' does not accept parameters, 'typing_extensions.NotRequired[int][int]' has 1.

I am sure I am overlooking something simple here, but I have everything I have tried doesn't seem to work and I have not been able to figure out the solution to this. I can tell this error is raised by [typeinfo](https://github.com/robotframework/robotframework/blob/master/src/robot/running/arguments/typeinfo.py) _report_nested_error, but I don't see any syntactical error with what i'm doing here. If I prevent this code from running, everything works like in RF 6

Does anyone have an idea of how I can get around this error? What am I missing here?

**Comment by pekkaklarck** [comment]
Could you create a simple runnable example I could test locally?

**Comment by gomarz** [comment]
Hi, here is a simple example
[Sample.zip](https://github.com/robotframework/robotframework/files/15019440/Sample.zip)
My error should be reproduceable by unzipping into a folder and running 

> robot Sample.robot

from within it

keywords.py
`from robot.api.deco import keyword
from signals import WriteSignal

@keyword(name="Write")
def write(self, *signals: WriteSignal) -> None:
    print("test")
    # self.test_library.publish_can_message("write", signals)
`
Sample.robot
`
*** Settings ***
Metadata         Version                                                                                                  0.1
Documentation    Simple example showcasing NotRequired error:

Force Tags       PreCheck

Library                          ${EXECDIR}/keywords.py

*** Keywords ***

WriteValue
    [Documentation]                 Create dictionary and publishes CAN:Write message
    ...                             with signal values
    [Arguments]                     ${signal}                                                    ${value}                                  ${interval}=${None}
    &{DICT} =                       Create Dictionary                                            Network=CAN5                              Signal=${signal}                           Value=${value}
    Run Keyword If                  $interval is not None                                        Set To Dictionary                         ${CAN_DICT}                                Interval=${interval}
    keywords.Write                  ${DICT}

*** Test Cases ***

Change Language from Radio to a Language supported by IHU (French)
    WriteValue              1   1
    Sleep                      2

`
signals.py
`
from typing import Any, Literal, Protocol, Union
from typing_extensions import NotRequired, TypedDict

class Signal(TypedDict):
    """Base Signal"""

    Network: str
    Signal: str


class ReadSignal(Signal):
    """Read Signal"""

    Mode: Literal["Once", "OnChange"]
    Interval: NotRequired[Union[int, None]]


class WriteSignal(Signal):
    """Write Signal"""

    Value: str
    Interval: NotRequired[Union[int, None]]


class GeneralLogger(Protocol):  # pylint: disable=too-few-public-methods
    """Definition of logger, which might be used as async in MessageSubscriber"""

    def log(self, level: int, msg: object, *args: Any, **kwargs: Any) -> None:  # type: ignore[misc]
        """Log message with provided level"""

`

**Comment by pekkaklarck** [comment]
Thanks for the example. It could still be simplified (for example, the library file could contain all need Python code), but I got the issue reproduced. Interestingly it doesn't happen with Python 3.11 or 3.12, though. Which Python version do you use?

**Comment by pekkaklarck** [comment]
I investigated this further and the problem is that `typing_extensions.NotRequired` isn't handled properly by the standard `typing` module with Python 3.10 and earlier. The issue can be demonstrated with the following example:
```python
from typing import get_type_hints, TypedDict
from typing_extensions import NotRequired

class X(TypedDict):
    x: int
    y: NotRequired[int]

print('hints:', get_type_hints(X))
print('required:', X.__required_keys__)
```
When the above is executed with Python 3.10 or older, the `NotRequired` annotation is not handled at all and you get this:
```
hints: {'x': <class 'int'>, 'y': typing_extensions.NotRequired[int]}
required: frozenset({'x', 'y'})
```
With Python 3.11 the result is, without changing the code, this:
```
hints: {'x': <class 'int'>, 'y': <class 'int'>}
required: frozenset({'x'})
```
The `NotRequired` annotation was added in Python 3.11, so it's not a surprise it's handled properly with that version automatically. With older versions we need to handle it ourselves.so that we replace `NotRequired[type]` in type hints with just `type` and also remove keys with such annotation from required keys.

Notice that if we change the imports in the above example so that we import everything from `typing_extensions`, `NotRequired` is handled properly. Because we cannot control how users import `TypedDict` and `NotRequired`, and because it feels a bit risky for us to use `get_type_hints` from `typing_extensions` instead of `typing`, I don't think that's a possible solution for the reported issue. Handling `NotRequired` ourselves is safer and also pretty simple.
